File: backend/app/sharekey.py
"""Share key — the credential the Android no-display share Activity carries.

⚠️ WHY THIS EXISTS AT ALL, because it looks like a second auth system and a
second auth system always deserves suspicion.

Android Phase B makes sharing invisible: a translucent no-display Activity
receives `ACTION_SEND`, saves the link and finishes, so the user never leaves
Instagram. That Activity runs **outside the JS runtime** — there is no React
Native context, no supabase-js — so it cannot use the Supabase session the way
every other request does. Two obvious routes were considered and rejected:

  Read the stored access token.  A Supabase access token lives ~1 h and only
      auto-refreshes while the app is open. Someone who opens Findable at
      breakfast and shares a reel at lunch has a four-hour-old token, so the
      silent save 401s. That is the MAJORITY case, not an edge case.

  Refresh the token from native code.  Supabase rotates refresh tokens, so a
      native refresh revokes the refresh token the app is still holding — and
      the next app launch signs the user out. Trading "a 1 s flash" for
      "randomly logged out" is a bad trade.

So the app mints a long-lived, **save-scoped** key while it holds a valid JWT,
and the Activity carries that instead.

SCOPE IS ENFORCED BY WHERE THIS IS ACCEPTED, NOT BY A CLAIM INSIDE THE KEY.
`user_for_share` is wired into exactly one route, `POST /api/reels/share-save`,
which is a three-line delegation to the normal save handler. Every other route
in the app still takes `get_current_user`. That is deliberately checkable by
reading the routing table: a stolen key can create a saved link and nothing
else — no read, no delete, no AI action, no account access. Giving it its own
URL rather than teaching `/save` to accept either credential is what keeps that
property a fact instead of a claim about a branch.

Stored as SHA-256, never in the clear — a database leak must not yield working
credentials. Revoked on sign-out, and it dies with the account automatically
because it lives on the profile row that account deletion already removes.

ONE KEY PER USER (latest device wins). Minting from a second device silently
supersedes the first, which then falls back to the visible launch path rather
than failing — acceptable while Android is the only platform with the Activity
(iOS is blocked on the Apple Developer account). If multi-device silent sharing
matters later this becomes its own table keyed by hash; nothing else changes.
"""
import hashlib
import secrets
from datetime import datetime

from fastapi import Depends, Header, HTTPException, status
from sqlalchemy.orm import Session

from app.auth import AuthUser
from app.database import ProfileDB, get_db
# The private helper on purpose: it also inherits the trial grant for a
# re-signed-up email. Creating a profile row without that would hand the user a
# fresh trial, which is the exact hole entitlements.py exists to close.
from app.entitlements import _get_or_create_profile, tier_for
from app.quota import quota_subject

#: Prefix so a leaked string is identifiable on sight in a log or bug report.
_PREFIX = "shk_"

#: The header the Activity sends. Deliberately NOT `Authorization` — a scheme
#: this narrow should not be mistakable for a session token by anything reading
#: a log or a proxy config.
SHARE_KEY_HEADER = "X-Share-Key"

_UNAUTHENTICATED = HTTPException(
    status_code=status.HTTP_401_UNAUTHORIZED,
    detail="Not authenticated",
    headers={"WWW-Authenticate": "Bearer"},
)


def _hash(key: str) -> str:
    return hashlib.sha256(key.encode("utf-8")).hexdigest()


def mint_share_key(db: Session, user: AuthUser) -> str:
    """Issue a fresh share key for `user`, replacing any previous one.

    Only ever called from a JWT-authenticated request, so the two snapshots
    below are taken from a VERIFIED token and never from client input.

    ⚠️ The snapshots are the whole reason this is more than one column. A
    share-key request carries no JWT, so:
      * `tier_for()` would find no `app_metadata` and read the caller as free —
        a paying Pro user's silent share would hit the free save cap;
      * `quota_subject()` would fall back to `user_id` instead of the
        normalized-email hash, charging a DIFFERENT daily AI bucket than the
        same person's normal saves. That is a quota hole, not a cosmetic one.

    They go stale between mints. The app re-mints on every launch, so the lag is
    the same class as the ~1 h JWT staleness the tier system already documents.
    """
    profile = _get_or_create_profile(db, user, now=datetime.utcnow())
    key = _PREFIX + secrets.token_urlsafe(32)
    profile.share_key_hash = _hash(key)
    profile.share_key_tier = tier_for(user)
    profile.share_key_subject = quota_subject(user)
    db.commit()
    return key


def revoke_share_key(db: Session, user: AuthUser) -> None:
    """Drop the user's share key (sign-out). Idempotent — a user who never
    minted one is not an error."""
    profile = db.query(ProfileDB).filter(ProfileDB.user_id == user.id).first()
    if profile is None or profile.share_key_hash is None:
        return
    profile.share_key_hash = None
    profile.share_key_tier = None
    profile.share_key_subject = None
    db.commit()


def resolve_share_key(db: Session, key: str) -> AuthUser | None:
    """The user this key belongs to, or None. Never raises."""
    if not key or not key.startswith(_PREFIX):
        return None
    profile = (
        db.query(ProfileDB).filter(ProfileDB.share_key_hash == _hash(key)).first()
    )
    if profile is None:
        return None
    # Rebuild the parts of the token the rest of the app reads. `email` stays
    # None deliberately — nothing downstream needs the address once `subject`
    # carries the identity the quota is charged against, and not storing it is
    # one less copy of a real email address at rest.
    return AuthUser(
        id=profile.user_id,
        email=None,
        claims={"app_metadata": {"tier": profile.share_key_tier or "free"}},
        subject=profile.share_key_subject,
    )


def user_for_share(
    share_key: str | None = Header(None, alias=SHARE_KEY_HEADER),
    db: Session = Depends(get_db),
) -> AuthUser:
    """The share-key caller, or 401. The ONLY dependency that accepts this
    credential, and it is attached to exactly one route."""
    user = resolve_share_key(db, share_key) if share_key else None
    if user is None:
        raise _UNAUTHENTICATED
    return user

File: mobile/services/shareKey.ts
import AsyncStorage from '@react-native-async-storage/async-storage';
import { Platform } from 'react-native';
import { api, apiBaseUrl } from './api';
import { setShareConfig, clearShareConfig } from '../modules/share-config';

/**
 * The credential the Android invisible-share Activity carries.
 *
 * ⚠️ WHY NOT JUST USE THE SUPABASE SESSION. The no-display share Activity runs
 * OUTSIDE the JS runtime — no React Native context, no supabase-js. Two obvious
 * routes were both rejected:
 *
 *   Read the stored access token.  It lives ~1 h and only auto-refreshes while
 *       the app is open, so someone who opened Findable at breakfast and shares
 *       a reel at lunch has a four-hour-old token. That is the MAJORITY case.
 *   Refresh it from Kotlin.  Supabase rotates refresh tokens, so a native
 *       refresh revokes the one the app is still holding and the next launch
 *       signs the user out. A 1 s flash is not worth a random logout.
 *
 * So the app mints a long-lived, save-scoped key while it holds a valid JWT and
 * parks it here. Backend rationale and scope proof: `backend/app/sharekey.py`.
 *
 * ⚠️ THE KEY AND THE SHAPE ARE A CONTRACT WITH KOTLIN *AND* SWIFT. AsyncStorage
 * on Android is the SQLite file `RKStorage` (table `catalystLocalStorage`), NOT
 * SharedPreferences — the Activity reads this exact row. `apiUrl` rides along
 * because `EXPO_PUBLIC_API_URL` is inlined into the JS bundle at build time and
 * simply does not exist natively; storing it here also means the Activity can
 * never talk to a different backend than the app.
 *
 * ⚠️ iOS NEEDS A SECOND COPY, IN A DIFFERENT PLACE. A Share Extension is its
 * own process with its own container and CANNOT read the app's AsyncStorage.
 * The App Group is the only surface both processes see, so the same JSON is
 * also written there through the `share-config` local module. Two stores, one
 * shape — if this object ever changes, both native readers change with it
 * (mobile/plugins/android/ShareSave.kt and
 * mobile/plugins/withInvisibleShareIOS.js).
 */

const KEY = '@savehere:sharekey:v1';
/** Did the LAST arming attempt actually finish? See `shareKeyReady`. */
const STATE = '@savehere:sharekey:armed:v1';

interface ShareConfig {
  key: string;
  apiUrl: string;
}

/**
 * True when the invisible share is armed on this device.
 *
 * ⚠️ THIS EXISTS BECAUSE THE FAILURE WAS INVISIBLE. Arming ran once per launch,
 * swallowed every error, and told nobody. A share then quietly fell back to
 * opening the app — indistinguishable, from the outside, from the bug the whole
 * extension was built to fix. There was no way for the owner to answer "is the
 * key even there?" and no way for me to ask their device.
 *
 * A cold Render free instance takes ~50 s to answer (see the note in
 * contexts/AuthContext.tsx), which is precisely how a launch-time mint loses.
 */
export async function shareKeyReady(): Promise<boolean> {
  if (Platform.OS === 'web') return false;
  return (await AsyncStorage.getItem(STATE).catch(() => null)) === 'armed';
}

/** Re-arm only if the last attempt didn't finish. Cheap to call on resume. */
export async function retryShareKeyIfNeeded(): Promise<void> {
  if (await shareKeyReady()) return;
  await ensureShareKey();
}

/**
 * Ensure a share key exists for this session. Safe to call on every launch —
 * minting is idempotent from the user's point of view (the new key replaces
 * the old one) and it is what keeps the server's tier/quota snapshots fresh.
 *
 * Best-effort by design: a failure here costs the *invisible* share, not the
 * share. Without a key the Activity forwards the intent to the app, which is
 * exactly today's Phase A behaviour.
 */
export async function ensureShareKey(): Promise<boolean> {
  // Both native platforms now have an invisible share: Android's ShareActivity
  // and, since 2026-09-10, the iOS Share Extension. Web has neither.
  if (Platform.OS === 'web') return false;

  // ⚠️ RETRIED, because ONE attempt at launch is the wrong number. Minting
  // needs the backend, the backend is a free Render instance, and a cold one
  // takes ~50 s to answer — so the request most likely to fail is the one fired
  // the instant the app opens. It used to fail there silently and never try
  // again until the next cold start, leaving the share falling back to opening
  // the app for the rest of the session.
  const BACKOFF = [0, 3000, 12000];
  for (let i = 0; i < BACKOFF.length; i++) {
    if (BACKOFF[i]) await new Promise(r => setTimeout(r, BACKOFF[i]));
    try {
      const { key } = await api.createShareKey();
      const config: ShareConfig = { key, apiUrl: apiBaseUrl() };
      const json = JSON.stringify(config);
      await AsyncStorage.setItem(KEY, json);
      // iOS reads from the App Group, not from AsyncStorage.
      //
      // ⚠️ THE RETURN VALUE IS CHECKED NOW. It was discarded, so a build whose
      // native module was missing or unentitled stored a perfectly good key in
      // a place the Share Extension cannot see and reported success. "We wrote
      // it" and "the other process can read it" are different claims.
      if (Platform.OS === 'ios' && !setShareConfig(json)) {
        await AsyncStorage.setItem(STATE, 'unarmed').catch(() => {});
        return false;   // native module absent — retrying cannot fix it
      }
      await AsyncStorage.setItem(STATE, 'armed').catch(() => {});
      return true;
    } catch {
      // Offline, cold backend, or a build that predates the endpoint. Leave
      // whatever is already stored — an older key still works until superseded.
    }
  }
  await AsyncStorage.setItem(STATE, 'unarmed').catch(() => {});
  return false;
}

/**
 * Drop the key on sign-out, locally AND on the server.
 *
 * The local delete is the one that matters for correctness (the Activity must
 * not keep saving into an account nobody is signed into on this device); the
 * server revoke is because a credential that outlives the session it was
 * minted from should be killable from the other end too.
 */
export async function clearShareKey(): Promise<void> {
  await AsyncStorage.removeItem(KEY).catch(() => {});
  await AsyncStorage.removeItem(STATE).catch(() => {});
  // ⚠️ The App Group copy has to go too. Miss it and the Share Extension keeps
  // a working credential after sign-out and carries on saving into an account
  // nobody is signed into on this device — the exact failure the local delete
  // above exists to prevent, one process over.
  if (Platform.OS === 'ios') clearShareConfig();
  try {
    await api.revokeShareKey();
  } catch {
    // Signing out must never fail because a revoke did. The local copy is
    // gone, which is what stops this device sharing silently.
  }
}
